storage/netstore.go

```diff
@@ -18,318 +18,289 @@ package storage
+const (
+	// capacity for the fetchers LRU cache
+	fetchersCapacity = 500000
 )
 
-type NetFetcher interface {
-	Request(hopCount uint8)
-	Offer(source *enode.ID)
+// Fetcher is a struct which maintains state of remote requests.
+// Fetchers are stored in fetchers map and signal to all interested parties if a given chunk is delivered
+// the mutex controls who closes the channel, and make sure we close the channel only once
+type Fetcher struct {
+	Delivered chan struct{} // when closed, it means that the chunk this Fetcher refers to is delivered
+
+	// it is possible for multiple actors to be delivering the same chunk,
+	// for example through syncing and through retrieve request. however we want the `Delivered` channel to be closed only
+	// once, even if we put the same chunk multiple times in the NetStore.
+	once sync.Once
+
+	CreatedAt time.Time // timestamp when the fetcher was created, used for metrics measuring lifetime of fetchers
+	CreatedBy string    // who created the fetcher - "request" or "syncing", used for metrics measuring lifecycle of fetchers
+
+	RequestedBySyncer bool // whether we have issued at least once a request through Offered/Wanted hashes flow
 }
 
-// NetStore is an extension of local storage
+// NewFetcher is a constructor for a Fetcher
+func NewFetcher() *Fetcher {
+	return &Fetcher{make(chan struct{}), sync.Once{}, time.Now(), "", false}
+}
+
+// SafeClose signals to interested parties (those waiting for a signal on fi.Delivered) that a chunk is delivered.
+// It closes the fi.Delivered channel through the sync.Once object, because it is possible for a chunk to be
+// delivered multiple times concurrently.
+func (fi *Fetcher) SafeClose() {
+	fi.once.Do(func() {
+		close(fi.Delivered)
+	})
+}
+
+type RemoteGetFunc func(ctx context.Context, req *Request, localID enode.ID) (*enode.ID, error)
+
+// NetStore is an extension of LocalStore
 // it implements the ChunkStore interface
-// on request it initiates remote cloud retrieval using a fetcher
-// fetchers are unique to a chunk and are stored in fetchers LRU memory cache
-// fetchFuncFactory is a factory object to create a fetch function for a specific chunk address
+// on request it initiates remote cloud retrieval
 type NetStore struct {
 	chunk.Store
-	mu                sync.Mutex
-	fetchers          *lru.Cache
-	NewNetFetcherFunc NewNetFetcherFunc
-	closeC            chan struct{}
+	localID      enode.ID // our local enode - used when issuing RetrieveRequests
+	fetchers     *lru.Cache
+	putMu        sync.Mutex
+	requestGroup singleflight.Group
+	RemoteGet    RemoteGetFunc
 }
 
-var fetcherTimeout = 2 * time.Minute // timeout to cancel the fetcher even if requests are coming in
+// NewNetStore creates a new NetStore using the provided chunk.Store and localID of the node.
+func NewNetStore(store chunk.Store, localID enode.ID) *NetStore {
+	fetchers, _ := lru.New(fetchersCapacity)
```

Comment:
We ignore the error, because we `know` that we are creating a LRU cache with a positive size.